Improved Logger

yanniknelson opened the pull request:
Took pull request #34 and applied requested changes

Removed the Log.h file as it only added transitive includes, is is meh practice, always good to exclude exactly what you need.
Also added RAII wrapper class around the logger and Get is now GetOrCreate (which will also Init if not already) to handle lifetime.

yanniknelson committed:
Improved Logger

Took pull request #34 and applied requested changes

thedevmystic commented:
Well ok, I'm currently off schedule due to my exams are near, around mid february. So expect delays. Thanks :)

yanniknelson committed:
Added Trace Verbosity Levels

yanniknelson committed:
Improved logging performance

So turns out spdlog is not as speedy as I would have liked, after some tinkering, increasing the thread pool from 2 to 4 with a larger queue size improved this, for application performance.

However spdlog ignores the flush level for flushing to stdout  and stderr (this is dumb). This results in the stdout lagging behind the program MASSIVELY when the program executes quickly (e.g. for running a brainfck program).

This would be easily solved if only it respected the flush level (or at least gave to option to)

yanniknelson commented:
Might be a good idea to move to the boost logging library instead of spdlog given my discovery in the latest commit [evidence](https://github.com/gabime/spdlog/blob/a49456f7f21fb802c1000fd79ef05911aa2101b7/include/spdlog/sinks/stdout_sinks-inl.h#L76)

thedevmystic reviewed (approved):
A thing is, if we are going to change spdlog to boost library. This would be a complete rewrite from scratch. Therefore, I didn't review it thoroughly as the logic is stale already and would be rewritten for boost afterwards.
